lib/global_uid/allocator.rb

```diff
@@ -0,0 +1,61 @@
+module GlobalUid
+  class Allocator
+    attr_reader :allocations, :max_size, :incrementing_by, :connection
+
+    def initialize(max_size: 5, incrementing_by:, connection:)
+      @allocations = []
+      @max_size = max_size
+      @incrementing_by = incrementing_by
+      @connection = connection
+      validate_connection_increment(connection)
+    end
+
+    def add(table)
+      identifier = connection.insert("REPLACE INTO #{table} (stub) VALUES ('a')")
+      increment!(identifier)
+    end
+
+    def add_many(table, count:)
+      increment_by = validate_connection_increment(connection)
+
+      start_id = connection.insert("REPLACE INTO #{table} (stub) VALUES " + (["('a')"] * count).join(','))
+      identifiers = start_id.step(start_id + (count-1) * increment_by, increment_by).to_a
+      identifiers.each { |identifier| increment!(identifier) }
+      identifiers
+    end
+
+    private
+
+    def valid_increment?
+      allocations[1..-1].all? do |identifier|
+        (identifier - allocations[0]) % incrementing_by == 0
+      end
+    end
+
+    def increment!(identifier)
```

Comment:
The name `increment!` makes it sound like it's going to increment the identifier itself, whereas what this really does is record and validate it. I think you used to have it `validate_increment!`? That might be clearer.

Reply:
Good point, thanks.

Reply:
I think `validate_allocation` makes sense, since it's an allocator. `valid_increment?` should also be changed to `valid_allocation?` since it's not dealing with the `auto_increment_increment`.

The two public methods could then be `allocate(table)` and `allocate_many(table, count:)`